openedx/openedx-platform
# [BD-04] Create CustomTagTemplateBlock, convert TranslateCustomTagDescriptor to XBlock and remove RawDescriptor

## Description

- Creates `CustomTagTemplateBlock` for use with `CustomTagBlock` instead of `RawDescriptor`.
- Converts `TranslateCustomTagDescriptor` to `TranslateCustomTagBlock`.
- In https://github.com/edx/edx-platform/pull/25955 `HiddenDescriptor` (which was a subclass of `RawDescriptor` with a custom `student_view()`) was converted to an XBlock. It is used as the `default_class` by the `CachingDescriptorSystem` classes. However `RawDescriptor` is still being used by `XMLModuleStore`. This has been replaced by `HiddenDescriptor` as well.
- Removes `RawDescriptor` since it is no longer used.

Part of [XModule to XBlock Conversion work](https://openedx.atlassian.net/wiki/spaces/AC/pages/1472790755/XModule+to+XBlock+Conversion). 

- Which edX user roles will this change impact? "Developer"

## Testing Instructions
* Checkout the `master` branch.
* Create a course and import this archive:
* In the course check that there are two units: the `Custom Tag Unit` with `html`, `custom_tag`, `image` and `book` XBlocks and the `Poll Unit` with a `poll_question` XBlock.
* Go to `http://localhost:18010/admin/xblock_django/xblockconfiguration/add/` and add an entry for `poll_question` with it set to NOT enabled.
* Visit the `Poll Unit` and verify that the following message is shown instead of the XBlock: "ERROR: "poll_question" is an unknown component type. This component will be hidden in LMS."
* Delete both the units from the Course Outline.
* Checkout this branch, do `pip -e common/lib/xmodule` in `make studio-shell` and then do `make lms-restart` and `make-studio-restart`.
* Import the archive again.
* Check that both the Units have the same content in Studio. The `poll_question` XBlock should have been imported in but the error message should show in its place: "ERROR: "poll_question" is an unknown component type. This component will be hidden in LMS."
* Check that both the Units are visible in LMS. Though in the `Poll Unit` the `poll_question` XBlock will be hidden.
* Export the course. Check that the `vertical` and `custom_tags` dir content is the same as before. However the `poll_question` dir should NOT be present (because that XBlock could not be exported).

## Screenshots

**Custom Tag Unit in Studio**

![image](https://user-images.githubusercontent.com/5305552/110826295-bfe44900-82b6-11eb-9c7e-b2c00a9ef8c6.png)

![image](https://user-images.githubusercontent.com/5305552/110826342-ca9ede00-82b6-11eb-948c-a2cef5f26b8f.png)

**Custom Tag Unit in LMS**

![image](https://user-images.githubusercontent.com/5305552/110826389-d8546380-82b6-11eb-886c-d2876692f20e.png)

**Poll Unit in Studio after Poll XBlock has been disabled**

![image](https://user-images.githubusercontent.com/5305552/110826072-88759c80-82b6-11eb-901b-53a3efc9f96b.png)

**Poll Unit in LMS after Poll XBlock has been disabled**

![image](https://user-images.githubusercontent.com/5305552/110826196-a511d480-82b6-11eb-8644-fe50e9ddf8e5.png)

## MRO Analysis

Descriptor MRO | Module MRO | Block MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.backcompat_module.TranslateCustomTagDescriptor'> | | <class 'xmodule.template_module.TranslateCustomTagBlock'> |
<class 'xmodule.x_module.XModuleDescriptor'> | | |
<class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> |
  | <class 'xmodule.x_module.XModule'> | |
  | <class 'xmodule.x_module.XModuleToXBlockMixin'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> |
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | | This is not needed because there are no views.
<class 'xmodule.x_module.ResourceTemplates'> | | | This is not needed because there are no templates.
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> |
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> |
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> |
